### PR TITLE
sql: fix row_number reduce elision optimization

### PR DESCRIPTION
**Pull request by mkysel**
This is a proposed fix for row_number aggregation. 

It works around #8963, by disabling the offending assertion.

### Motivation

ROW_NUMBER would crash the server if reduce elision optimization takes place.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR adds a release note for any
  [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

File: src/expr/src/scalar/func.rs
Comment (umanwizard) [inline src/expr/src/scalar/func.rs:5754]: Can you please explain this more -- why is this assert firing? Why is it okay to ignore it?

#9008 doesn't seem to have any description.

Comment (mkysel) [reply src/expr/src/scalar/func.rs:5754]: I will have to defer to @asenac as he is working on it actively.

Comment (asenac) [reply src/expr/src/scalar/func.rs:5754]: It doesn't have a description because it wasn't meant to be shared yet. The description of the problem is in https://github.com/MaterializeInc/materialize/issues/8963.

File: src/expr/src/scalar/func.rs
Comment (asenac) [inline src/expr/src/scalar/func.rs:5755]: Please don't reference the pull request but the issue

**Review by umanwizard — APPROVED** [review]
I'm fine with this to unblock main, but we should follow up and fix this assert ASAP.

File: src/expr/src/scalar/func.rs
Comment (mkysel) [reply src/expr/src/scalar/func.rs:5755]: fixed

File: src/expr/src/scalar/func.rs
Comment (asenac) [inline src/expr/src/scalar/func.rs]: I think this was readded by accidentally

**Review by asenac — APPROVED** [review]
I'll put the assert back as part of the fix for the underlying issue

**Pull request merged**
